umbral/pre.py

```diff
@@ -327,32 +312,24 @@ def generate_kfrags(delegating_privkey: UmbralPrivateKey,
 
 
 def reencrypt(kfrag: KFrag,
-              capsule: Capsule,
-              provide_proof: bool = True,
+              prepared_capsule: PreparedCapsule,
               metadata: Optional[bytes] = None,
               verify_kfrag: bool = True) -> CapsuleFrag:
 
-    if not isinstance(capsule, Capsule) or not capsule.verify():
+    if not isinstance(prepared_capsule, PreparedCapsule):
+        raise Capsule.NotValid
+
+    if not prepared_capsule.verify():
         raise Capsule.NotValid
 
     if verify_kfrag:
-        if not isinstance(kfrag, KFrag) or not kfrag.verify_for_capsule(capsule):
+        if not isinstance(kfrag, KFrag) or not prepared_capsule.verify_kfrag(kfrag):
             raise KFrag.NotValid
 
-    rk = kfrag.bn_key
-    e1 = rk * capsule.point_e  # type: Any
-    v1 = rk * capsule.point_v  # type: Any
-
-    cfrag = CapsuleFrag(point_e1=e1, point_v1=v1, kfrag_id=kfrag.id,
-                        point_precursor=kfrag.point_precursor)
-
-    if provide_proof:
-        cfrag.prove_correctness(capsule, kfrag, metadata)
-
-    return cfrag
+    return CapsuleFrag.from_kfrag(prepared_capsule.capsule, kfrag, metadata)
```

Comment:
I'm not very convinced of this change. Re-encryption, the current cryptographic core of the NuCypher network gets demoted behind `CapsuleFrag.from_kfrag()`. On second thoughts, maybe I'm just being dramatic here, but I'd like others' input here too.

Reply:
Also not a big fan of this change. If the core of the API is central to Proxy Re-Encryption, then hiding where it happens in a method like this makes it a bit difficult to understand and use.

Reply:
The user still only sees `reencrypt()`. Would it be better to have `KFrag.reencrypt() -> CFrag` instead or `CapsuleFrag.from_kfrag()`?

Reply:
Or even `Capsule.reencrypt(kfrag) -> CFrag`

Reply:
What's wrong with the original place?

Reply:
The rationale for putting this code in `from_kfrag()` is that it is the only way to create a `CapsuleFrag` (besides deserialization), so it is logical to make it a (classmethod) constructor. Both to ensure that a `CapsuleFrag` is always in a valid state, and that all the ways to create it are easily discoverable. 

A somewhat weaker argument is that it is not common in Python to have a free-standing function accessing internal state of objects of several types. If I were looking for a way to create a `CapsuleFrag`, I'd look into methods of `CapsuleFrag`, `KFrag` and `Capsule` first before searching for a separate function.

That said, this change is not essential to the main purpose of the PR, and I can roll it back if you're opposed to it.

Reply:
All those are valid arguments. I'd like the opinion of @jMyles and @KPrasch too. At some moment, we decided to leave the core logic in the `pre.py` module as standalone functions, instead of encapsulating them in any class. But this was 2.5 years ago and it's rained a lot since then.

Reply:
`pre.reencrypt()` isn't really going anywhere, it just becomes a thin wrapper over `from_kfrag()` or wherever the logic ends up.